src/Mvc/Mvc.Core/src/Routing/UrlHelperBase.cs

```diff
@@ -43,45 +43,43 @@ protected UrlHelperBase(ActionContext actionContext)
         /// <inheritdoc />
         public virtual bool IsLocalUrl(string url)
         {
-            if (string.IsNullOrEmpty(url))
+            ReadOnlySpan<char> span = url;
+
+            if (span.IsEmpty)
             {
                 return false;
             }
 
-            // Allows "/" or "/foo" but not "//" or "/\".
-            if (url[0] == '/')
+            if (span[0] == '~')
             {
-                // url is exactly "/"
-                if (url.Length == 1)
-                {
-                    return true;
-                }
-
-                // url doesn't start with "//" or "/\"
-                if (url[1] != '/' && url[1] != '\\')
-                {
-                    return true;
-                }
+                span = span[1..];
+            }
 
+            // url does not start with "/" or "~/"
+            if (span[0] != '/')
+            {
                 return false;
             }
 
-            // Allows "~/" or "~/foo" but not "~//" or "~/\".
-            if (url[0] == '~' && url.Length > 1 && url[1] == '/')
+            // url is exactly "/" or "~/"
+            if ((uint)span.Length <= 1)     // PERF: elides the bounds-check below
             {
-                // url is exactly "~/"
-                if (url.Length == 2)
-                {
-                    return true;
-                }
+                return true;
+            }
 
-                // url doesn't start with "~//" or "~/\"
-                if (url[2] != '/' && url[2] != '\\')
+            // url doesn't start with "//" or "/\"
+            if (span[1] != '/' && span[1] != '\\')
+            {
+                // check that url does not contain any (ASCII) control character
+                for (var i = 1; i < span.Length; ++i)
                 {
-                    return true;
+                    if ((uint)(span[i] - 0x20) >= 0x7F - 0x20)  // span[i] < 0x20 || span[i] == 0x7F
```

Comment:
This check is only safe under the assumption of ASCII. Maybe I should add an assert for this.
Otherwise I can revert https://github.com/aspnet/AspNetCore/commit/e4f6bc377a118e90bad8b2440005ad058891723e

Reply:
Stick with slow and correct. You can't make any assumptions/assertions about the input.

Reply:
I've reverted the last commit.